keystore/src/connection/platform/wasm/migrations.rs

```diff
@@ -0,0 +1,416 @@
+
+            // See docstring. Here, alternatively, we'd use the identifiers constructed above for
+            // each entity or use tuple indexing once one of the required language features is
+            // stable.
+            let ( $( $conversion_result, )* ) = conversion_results;
+
+            $(
+                match $conversion_result {
+                    Ok(mut collection) => storage_during_migration.save(<$entity>::COLLECTION_NAME, &mut collection).await?,
+                    // If this error occurs, we assume (for now) that it is due to the old aad not
+                    // working because the entity has already been migrated.
+                    // So we ignore this to recover from partial migrations.
+                    Err(CryptoKeystoreError::CryptoKeystoreErrorV1_0_0(CryptoKeystoreErrorV1_0_0::AesGcmError)) => {},
+                    Err(e) => Err(e)?,
+                };
+            )*
+
+            // Check for each entity whether the migration was successful, i.e., all records in the
+            // database can be decrypted.
+            let is_migration_successful = vec![$(
+                <$entity>::was_migrated_to_db_version_1(&storage_during_migration).await,
+            )*].iter().all(|&x| x);
+
+            storage_during_migration.close()?;
+
+            // If the migration wasn't successful, we cannot recover
+            if !is_migration_successful {
+                return Err(CryptoKeystoreError::MigrationFailed)
+            }
+
+            // Otherwise the migration is complete and the version counter can be incremented, which
+            // will happen at the end of open_and_migrate().
+            Ok(db_version_number(1))
+        }
+    };
+}
+
+/// Migrates from any old DB version to DB version 1.
+///
+/// _**Assumption**_: the entire storage fits into memory
+async fn migrate_to_version_1(name: &str, key: &str) -> CryptoKeystoreResult<u32> {
+    migrate_entities_to_version_1!(
+        name,
+        key,
+        [
+            (identifier_01, MlsCredential),
+            (identifier_02, MlsSignatureKeyPair),
+            (identifier_03, MlsHpkePrivateKey),
+            (identifier_04, MlsEncryptionKeyPair),
+            (identifier_05, MlsEpochEncryptionKeyPair),
+            (identifier_06, MlsPskBundle),
+            (identifier_07, MlsKeyPackage),
+            (identifier_08, PersistedMlsGroup),
+            (identifier_09, PersistedMlsPendingGroup),
+            (identifier_10, MlsPendingMessage),
+            (identifier_11, E2eiEnrollment),
+            (identifier_12, E2eiRefreshToken),
+            (identifier_13, E2eiAcmeCA),
+            (identifier_14, E2eiIntermediateCert),
+            (identifier_15, E2eiCrl),
+            (identifier_16, ProteusPrekey),
+            (identifier_17, ProteusIdentity),
+            (identifier_18, ProteusSession)
+        ]
+    )
+}
+
+fn get_builder(name: &str, version: u32) -> DatabaseBuilder {
```

Comment:
As I see it, `get_builder` defines the schema for the db. As such, I'm wondering whether the argument `version` makes sense.

For example, consider that we want to add another db version (say, 2), and a migration for it.
Suppose that version 2 replaces 1 db store with 2 new stores. Without embedding extra info about the new db version into `get_builder`, we cannot just call `get_builder(..., 2)`
because that would return an incorrect schema.

I suggest to have the version hardcoded with the builder. For instance,

```rust
fn get_builder_v0(name: &str) -> DatabaseBuilder {
    let idb_builder = DatabaseBuilder::new(name)
        .version(0)
    // a bunch of add_object_store calls
}

fn get_builder_v1(name: &str) -> DatabaseBuilder {
    let idb_builder = DatabaseBuilder::new(name)
        .version(1)
    // a bunch of add_object_store calls
}

fn get_builder_v2(name: &str) -> DatabaseBuilder {
    let idb_builder = DatabaseBuilder::new(name)
        .version(2)
    // a bunch of add_object_store calls
}
```
Now, one could reduce duplication by factoring out "a bunch of add_object_store calls" into some utility functions,  though we would need to be extra-careful when making changes to those utility functions in order to avoid breaking old migrations.

In cases you have a migration that merely adds another store, one could simply do
```rust
fn get_builder_v3(name: &str) -> DatabaseBuilder {
    let idb_builder = get_builder_v2(name).
        .version(3)
        .add_object_store(...)
}
```